src/libs/maintaining-std.md

```diff
@@ -0,0 +1,132 @@
+
+New unstable features don’t need an RFC before they can be merged. They do need one before they can be stabilized. New impls for already stable traits don't need an RFC. If a new unstable feature is large or has a lot of design space to explore then you might want to block merging it on an RFC.
+
+### Is there any `unsafe`?
+
+Unsafe code blocks in the standard library need a comment explaining why they're [ok](https://doc.rust-lang.org/nomicon). There's a `tidy` lint that checks this. The unsafe code also needs to actually be ok.
+
+The rules around what's sound and what's not can be subtle. See the [Unsafe Code Guidelines WG] for current thinking, and consider pinging `@rust-lang/libs` and/or somebody from the WG if you're in any doubt.
+
+### Is that `#[inline]` right?
+
+Inlining is a trade-off between potential execution speed, compile time and code size.
+
+You should add `#[inline]`:
+
+- To public, small, non-generic functions.
+
+You shouldn’t need `#[inline]`:
+
+- On methods that have any generics in scope.
+- On methods on traits that don’t have a default implementation.
+- On `const` items.
+
+#### What about `#[inline(always)]`?
+
+You should just about never need `#[inline(always)]`. It may be beneficial for private helper methods that are used in a limited number of places or for trivial operators. A micro benchmark should justify the attribute.
+
+### Is there any potential breakage?
+
+Breaking changes should be avoided when possible. [RFC 1105] lays the foundations for what constitutes a breaking change. Breakage may be deemed acceptable or not based on its actual impact, which can be approximated with a `crater` run.
+
+For changes where the value is high and the impact is high too, there are strategies for minimizing the impact:
+
+- Using compiler lints to try phase out broken behavior.
+
+### Could this affect inference?
+
+New impls on public traits for public items may cause inference to break when there are generics involved.
+
+#### Are there `#[fundamental]` items involved?
+
+Blanket trait impls can't be added to `#[fundamental]` types because they have different coherence rules. That includes:
+
+- `Box<T>`
+- `Pin<T>`
+
+Also see [RFC 1023] for details.
+
+### Is there specialization involved?
+
+We try to avoid leaning on specialization too heavily, limiting its use to optimizing specific implementations. Any use of specialization that changes how methods are dispatched for external callers should be carefully considered.
+
+### Does this change drop order?
+
+Changes to collection internals may affect the order their items are dropped in. This has been accepted in the past, but should be noted.
+
+### Could `mem::replace` break assumptions?
+
+Any value behind a `&mut` reference can be replaced with a new one.
+
+### Could `mem::forget` break assumptions?
+
+Rust doesn't guarantee destructors will run, so code should avoid relying on them for safety. Remember, [everybody poops][Everybody Poops].
```

Comment:
cc @rust-lang/wg-unsafe-code-guidelines

Reply:
Seems accurate in general. However, also see [these UCG issues](https://github.com/rust-lang/unsafe-code-guidelines/issues?q=is%3Aissue+is%3Aopen+label%3AT-drop). And note that for *some* instances, we *do* guarantee things about destructors -- I am talking about the [drop guarantee for pinned data](https://doc.rust-lang.org/nightly/std/pin/index.html#drop-guarantee). So, just because `mem::forget` is safe doesn't mean you can just drop arbitrary storage (like stack frames) without running the destructors. `mem::forget` *leaks*, that is okay -- but it doesn't *deallocate without dropping*, which is not (in general) okay.

Reply:
Ah thanks, I wasn't familiar with the drop guarantee for `Pin`!

The kind of undesirable invariant I was trying to catch with this section was just along the same lines as `Vec::drain`. Maybe we should reword this one around whether or not there are manual impls for `Drop`, and some of the things to keep in mind there? Like remembering that leaking is a thing?

Reply:
It is definitely worth mentioning that `mem::forget` is a thing, and that old `Vec::drain` and `thread::spawn` are unsound. But at the same time I think we should remind the reader that *everything can leak* does not imply that we can just *deallocate any storage without calling destructors*. This is an incorrect inference I have seen several people make.

Reply:
So what we're saying here is that, in general, you can't guarantee an impl of `Drop` will get called, but you must ensure that an impl of `Drop` _will_ get called if its storage is being deallocated/repourposed through something like `ptr::write`? And that's on the burden of the caller to `ptr::write` to enforce?

Reply:
Yes. Calling all relevant `Drop` impls is a *precondition* to deallocating or repurposing storage.

Or, in other words: there is no guarantee that an impl of `Drop` will get called. However, *if* the impl of `Drop` does not get called, then the memory backing that instance of the type must leak.